File: src/platform/FogPlatform.cpp
#include "platform/FogPlatform.hpp"
#include <sstream>

namespace enigma {

ZoneConfig FogPlatform::createHierarchicalTopology(int numFogNodes,
                                                    const std::string& nodeSpeed) {
    ZoneConfig zone("fog_hierarchical", "Full");
    
    // Create fog nodes in hierarchy
    for (int i = 0; i < numFogNodes; ++i) {
        std::string id = "fog_node_" + std::to_string(i);
        zone.hosts.emplace_back(id, nodeSpeed, 8);
    }
    
    // Links between fog nodes (high capacity)
    for (int i = 0; i < numFogNodes; ++i) {
        for (int j = i + 1; j < numFogNodes; ++j) {
            std::string linkId = "fog_link_" + std::to_string(i) + "_" + std::to_string(j);
            zone.links.emplace_back(linkId, "1GBps", "5ms");
        }
    }
    
    return zone;
}

ZoneConfig FogPlatform::createEdgeFogTopology(int numFogNodes, int edgeDevicesPerNode,
                                               const std::string& fogSpeed,
                                               const std::string& edgeSpeed) {
    ZoneConfig zone("edge_fog_topology", "Full");
    
    // Create fog nodes
    for (int i = 0; i < numFogNodes; ++i) {
        std::string fogId = "fog_node_" + std::to_string(i);
        zone.hosts.emplace_back(fogId, fogSpeed, 8);
        
        // Create edge subzone for each fog node
        ZoneConfig edgeSubzone("edge_zone_" + std::to_string(i), "Full");
        
        for (int j = 0; j < edgeDevicesPerNode; ++j) {
            std::string edgeId = "edge_" + std::to_string(i) + "_" + std::to_string(j);
            edgeSubzone.hosts.emplace_back(edgeId, edgeSpeed, 1);
        }
        
        // Link between edge and fog
        edgeSubzone.links.emplace_back("edge_to_fog_link", "500MBps", "8ms");
        
        zone.subzones.push_back(edgeSubzone);
    }
    
    // Links between fog nodes
    zone.links.emplace_back("fog_interconnect", "1GBps", "3ms");
    
    return zone;
}

ZoneConfig FogPlatform::createGeographicTopology(int numRegions, int nodesPerRegion,
                                                  const std::string& nodeSpeed) {
    ZoneConfig zone("fog_geographic", "Full");
    
    // Create geographic regions
    for (int r = 0; r < numRegions; ++r) {
        ZoneConfig region("region_" + std::to_string(r), "Full");
        
        // Fog nodes in each region
        for (int n = 0; n < nodesPerRegion; ++n) {
            std::string id = "fog_r" + std::to_string(r) + "_n" + std::to_string(n);
            region.hosts.emplace_back(id, nodeSpeed, 8);
        }
        
        // Intra-region links (low latency)
        region.links.emplace_back("intra_region_link", "1GBps", "2ms");
        
        zone.subzones.push_back(region);
    }
    
    // Inter-region links (higher latency)
    zone.links.emplace_back("inter_region_link", "500MBps", "20ms");
    
    return zone;
}

HostConfig FogPlatform::createFogNode(const std::string& id,
                                      const std::string& nodeType) {
    if (nodeType == "lightweight") {
        return HostConfig(id, "5Gf", 4);
    } else if (nodeType == "standard") {
        return HostConfig(id, "10Gf", 8);
    } else if (nodeType == "powerful") {
        return HostConfig(id, "20Gf", 16);
    } else if (nodeType == "edge_server") {
        return HostConfig(id, "15Gf", 12);
    } else {
        return HostConfig(id, "10Gf", 8);
    }
}

LinkConfig FogPlatform::createFogLink(const std::string& id,
                                      const std::string& linkType) {
    if (linkType == "ethernet") {
        return LinkConfig(id, "1GBps", "1ms");
    } else if (linkType == "fiber") {
        return LinkConfig(id, "10GBps", "500us");
    } else if (linkType == "wireless_5g") {
        return LinkConfig(id, "2GBps", "3ms");
    } else if (linkType == "wan") {
        return LinkConfig(id, "500MBps", "15ms");
    } else {
        // Default ethernet
        return LinkConfig(id, "1GBps", "2ms");
    }
}

} // namespace enigma

File: src/python/enigma/platform/fog_platform.py
"""
FogPlatform: specialized factories for fog computing platform topologies.
Mirrors C++ FogPlatform class from src/platform/FogPlatform.cpp.
"""

from __future__ import annotations

from .configs import HostConfig, LinkConfig, ZoneConfig


class FogPlatform:
    """Factory for fog computing platform topologies."""

    # ------------------------------------------------------------------ #
    # Topology factories                                                   #
    # ------------------------------------------------------------------ #

    @staticmethod
    def create_hierarchical_topology(
        num_fog_nodes: int,
        node_speed: str = "10Gf",
    ) -> ZoneConfig:
        """Flat hierarchy of interconnected fog nodes."""
        zone = ZoneConfig("fog_hierarchical", "Full")
        for i in range(num_fog_nodes):
            zone.hosts.append(HostConfig(f"fog_node_{i}", node_speed, 8))
        for i in range(num_fog_nodes):
            for j in range(i + 1, num_fog_nodes):
                zone.links.append(LinkConfig(f"fog_link_{i}_{j}", "1GBps", "5ms"))
        return zone

    @staticmethod
    def create_edge_fog_topology(
        num_fog_nodes: int,
        edge_devices_per_node: int,
        fog_speed: str = "10Gf",
        edge_speed: str = "1Gf",
    ) -> ZoneConfig:
        """
        Hierarchical edge-fog topology:
        each fog node owns a sub-zone of edge devices.
        """
        zone = ZoneConfig("edge_fog_topology", "Full")
        for i in range(num_fog_nodes):
            zone.hosts.append(HostConfig(f"fog_node_{i}", fog_speed, 8))
            sub = ZoneConfig(f"edge_zone_{i}", "Full")
            for j in range(edge_devices_per_node):
                sub.hosts.append(HostConfig(f"edge_{i}_{j}", edge_speed, 1))
            sub.links.append(LinkConfig("edge_to_fog_link", "500MBps", "8ms"))
            zone.subzones.append(sub)
        zone.links.append(LinkConfig("fog_interconnect", "1GBps", "3ms"))
        return zone

    @staticmethod
    def create_geographic_topology(
        num_regions: int,
        nodes_per_region: int,
        node_speed: str = "10Gf",
    ) -> ZoneConfig:
        """Geographic regions, each with their own set of fog nodes."""
        zone = ZoneConfig("fog_geographic", "Full")
        for r in range(num_regions):
            region = ZoneConfig(f"region_{r}", "Full")
            for n in range(nodes_per_region):
                region.hosts.append(HostConfig(f"fog_r{r}_n{n}", node_speed, 8))
            region.links.append(LinkConfig("intra_region_link", "1GBps", "2ms"))
            zone.subzones.append(region)
        zone.links.append(LinkConfig("inter_region_link", "500MBps", "20ms"))
        return zone

    # ------------------------------------------------------------------ #
    # Individual component factories                                       #
    # ------------------------------------------------------------------ #

    @staticmethod
    def create_fog_node(node_id: str, node_type: str = "standard") -> HostConfig:
        presets = {
            "lightweight": HostConfig(node_id, "5Gf",  4),
            "standard":    HostConfig(node_id, "10Gf", 8),
            "powerful":    HostConfig(node_id, "20Gf", 16),
            "edge_server": HostConfig(node_id, "15Gf", 12),
        }
        return presets.get(node_type, HostConfig(node_id, "10Gf", 8))

    @staticmethod
    def create_fog_link(link_id: str, link_type: str = "ethernet") -> LinkConfig:
        presets = {
            "ethernet":  LinkConfig(link_id, "1GBps",  "2ms"),
            "fiber":     LinkConfig(link_id, "10GBps", "500us"),
            "wireless":  LinkConfig(link_id, "300MBps","5ms"),
            "wan":       LinkConfig(link_id, "100MBps","20ms"),
        }
        return presets.get(link_type, LinkConfig(link_id, "1GBps", "2ms"))
